src/colmap/estimators/cost_functions.h

```diff
@@ -453,6 +453,47 @@ struct MetricRelativePoseErrorCostFunction {
   const EigenMatrix6d sqrt_information_j_;
 };
 
+// Cost function for aligning one 3D point with a reference 3D point with
+// covariance r = s * R * p + t - p_ref
+struct Point3dAlignmentCostFunction {
+ public:
+  Point3dAlignmentCostFunction(const Eigen::Vector3d& ref_point,
+                               const Eigen::Matrix3d& covariance_point)
+      : ref_point_(ref_point),
+        sqrt_information_point_(SqrtInformation(covariance_point)) {}
+
+  static ceres::CostFunction* Create(const Eigen::Vector3d& ref_point,
+                                     const Eigen::Matrix3d& covariance_point) {
+    return (
+        new ceres::
+            AutoDiffCostFunction<Point3dAlignmentCostFunction, 3, 3, 4, 3, 1>(
+                new Point3dAlignmentCostFunction(ref_point, covariance_point)));
+  }
+
+  template <typename T>
+  bool operator()(const T* const point,
+                  const T* const transform_q,
+                  const T* const transform_t,
+                  const T* const log_scale,
```

Comment:
What would it take to be able to directly pass fields of `Sim3d` object? For example:
```python
tfm = pycolmap.Sim3d()
prob.add_residual_block(pycolmap.Point3dAlignmentCost(...), loss, [pt, tfm.rotation, tfm.translation, tfm.scale])
```
1) A custom manifold parameterization that enforces a strictly positive scale?
2) `Sim3d.scale` be a numpy array instead of a double?

Reply:
While this has just been done in the new commit, there is no easy way to bind it to Python since we dont have access to the base binding of PyManifold here: https://github.com/cvg/pyceres/blob/main/_pyceres/core/manifold.h#L62

Reply:
> 2. `Sim3d.scale` be a numpy array instead of a double?

This is done by adding a new interface ``scale_ptr``. Now we can do
```
prob.add_residual_block(cost, loss, [pt, tfm.rotation, tfm.translation, tfm.scale_ptr])
```
The only thing that needs to be resolved is the binding of manifold. The easy way is to add that manifold to pyceres.

Reply:
Try this out
```c++
      .def_property(
          "scale",
          [](Sim3d& self) { return py::array({}, {}, &self.scale, py::cast(self)); },
          [](Sim3d& self, double scale) { self.scale = scale; })
```

Reply:
Regarding the binding of the manifold object: all bindings in pyceres (SubsetManifold, etc.) only reference the base `ceres::Manifold`, not PyManifold. Shouldn't this work here too if pyceres has been imported? does it then fail if pyceres hasn't been imported?

Reply:
We could provide such bindings only if pyceres is available:
```c++
bool pyceres_is_available = true;
try {
  py::module_::import("pyceres");
} catch (const py::import_error&) {
  pyceres_is_available = false;
}
if (pyceres_is_available) {
  // bind custom manifold
}
```

Reply:
> Try this out
> 
> ```c++
>       .def_property(
>           "scale",
>           [](Sim3d& self) { return py::array({}, {}, &self.scale, py::cast(self)); },
>           [](Sim3d& self, double scale) { self.scale = scale; })
> ```

Thanks. This works in my test.

Reply:
> We could provide such bindings only if pyceres is available:
> 
> ```c++
> bool pyceres_is_available = true;
> try {
>   py::module_::import("pyceres");
> } catch (const py::import_error&) {
>   pyceres_is_available = false;
> }
> if (pyceres_is_available) {
>   // bind custom manifold
> }
> ```

While this could work, this has one clear drawback. Assume the users have pyceres installed and this part of code is thus binded, they will be forced to import pyceres before importing pycolmap. Otherwise the import of pycolmap will raise an error because the ceres::Manifold binding is not found. This does not feel ideal to the pycolmap package.

Reply:
> > We could provide such bindings only if pyceres is available:
> > ```c++
> > bool pyceres_is_available = true;
> > try {
> >   py::module_::import("pyceres");
> > } catch (const py::import_error&) {
> >   pyceres_is_available = false;
> > }
> > if (pyceres_is_available) {
> >   // bind custom manifold
> > }
> > ```
> 
> While this could work, this has one clear drawback. Assume the users have pyceres installed and this part of code is thus binded, they will be forced to import pyceres before importing pycolmap. Otherwise the import of pycolmap will raise an error because the ceres::Manifold binding is not found. This does not feel ideal to the pycolmap package.

Also, seems that the CI does not accept try catch on the py::import_error: https://github.com/colmap/colmap/actions/runs/9727459725/job/26847035961?pr=2621

Reply:
> We could provide such bindings only if pyceres is available:
> 
> ```c++
> bool pyceres_is_available = true;
> try {
>   py::module_::import("pyceres");
> } catch (const py::import_error&) {
>   pyceres_is_available = false;
> }
> if (pyceres_is_available) {
>   // bind custom manifold
> }
> ```

py::import_error cannot be caught when pyceres is not installed. Changing it to the general py::error_already_set resolves the issue.